accepted/2022/simplify-output-paths.md

```diff
@@ -0,0 +1,53 @@
+
+- `bin\<Configuration>\<TargetFramework>` - Build output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>` - Build output with `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\publish` - Publish output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>\publish` - Publish output with `RuntimeIdentifier`
+
+This makes it harder to find the output of a project, and sometimes nests one type of output in another output folder.
+
+## Proposed behavior
+
+Projects targeting .NET 8 and higher will by default not include the `TargetFramework` in the output path unless the project is multi-targeted.  Similarly, the `RuntimeIdentifier` will not by default be part of the output path.  Finally, the publish output path will be `bin\publish` by default instead of a `publish` subfolder inside of the output path.
+
+These would be the default output paths for .NET 8 projects:
+
+- `bin\<Configuration>` - Build output path for single targeted project
+- `bin\<Configuration>\<TargetFramework>` - Build output for multi-targeted project
+- `bin\publish` - Publish output for single targeted project
+- `bin\publish\<TargetFramework>` - Publish output for multi-targeted project
+
+## Implementation
+
+There are currently properties that control whether the TargetFramework and RuntimeIdentifier should be included in the output path: `AppendTargetFrameworkToOutputPath` and `AppendRuntimeIdentifierToOutputPath`.  Both of these currently default to true.  We will change the default values as follows:
+
+- `AppendTargetFrameworkToOutputPath` will default to true if the `TargetFrameworks` property is set (as multi-targeted projects still need different output paths for each TargetFramework), or the project is targeting .NET 7 or lower.  Otherwise it will default to false.
+- `AppendRuntimeIdentifierToOutputPath` will default to true only when targeting .NET 7 or lower.
+  - To be discussed: Should we also default it to true if `RuntimeIdentifiers` is set?
+
+For publish, we would change the publish output path only when targeting .NET 8 or higher and when the Configuration is set to Release (which we [plan to make the default](https://github.com/dotnet/sdk/issues/10357#issuecomment-1208564539) for .NET 8).  In that case we would set the publish output path to `$(BaseOutputPath)\publish` (by default `bin\publish`), and append the `TargetFramework` and `RuntimeIdentifier` to the path based on `AppendTargetFrameworkToOutputPath` and `AppendRuntimeIdentifierToOutputPath`.
+
+## Considerations
+
+### Breaking changes
+
+These changes could break things such as scripts that copy the output of the build or custom MSBuild logic that hard-codes these paths.  Tieing these changes to the project TargetFramework ensures that these breaks will be encountered when the project is modified to target a new TargetFramework, not when updating to a new version of the .NET SDK.
```

Comment:
Is there a way, or can we add a way, to get the paths, so scripts can programatically get these directories rather than hardcode them?  For example `$publishPath = & dotnet publish -f net8.0 -r linux-x64 --output-path`?  Is there already a target that we can use, something like `$publishPath = & dotnet msbuild -t:GetPublishPath -p:TargetFramework=net8.0 -p:RuntimeIdentifer=linux-x64`?

If msbuild doesn't already have a feature to output the value of a single property after an evaluation (and running a target?), maybe that would be a useful feature?  `$publishPath = & dotnet msbuild --output-property PublishDir -p:TargetFramework=net8.0 -p:RuntimeIdentifer=linux-x64`

Reply:
I don't think there's a direct way for MSBuild to return a value to a script that calls it.

Targets can return properties or items, and this is used by Visual Studio to get information from the build.  Maybe it would be an interesting feature to support some way of doing that from the command line (that would be separate from this proposal though).  @rainersigwald, what do you think?

Reply:
dotnet/msbuild#3911 tracks some options there, which I'm not opposed to but I don't see a nice elegant solution for it at the moment, so it hasn't gotten a ton of traction. Feedback welcome!